cloudferrylib/os/network/neutron.py

```diff
@@ -1087,52 +1085,71 @@ def upload_networks(self, networks):
-                # attribute to avoid seg_id overlap)
+                # Update network info with additional arguments.
+                # We need to check if we have parameter
+                # "provider:physical_network" or param
+                # "provider:network_type" either is 'gre' or 'vxlan'.
+                # If condition is satisfied, we need to specify 2 more params:
+                # "provider:network_type" and "provider:segmentation_id".
                 list_update_atr = ["provider:network_type"]
                 if phys_net:
                     list_update_atr.append("provider:physical_network")
                 for atr in list_update_atr:
                     network_info['network'].update({atr: src_net.get(atr)})
 
-                if src_net["provider:segmentation_id"]:
-                    network_info['network'][
-                        'provider:segmentation_id'] = src_net[
-                        'provider:segmentation_id']
-                self.create_network(src_net, network_info)
-            else:
-                # create networks later (to be sure that generated
-                # segmentation ids don't overlap segmentation ids
-                # created manually)
-                networks_without_seg_ids.update({
-                    src_net.get("id"): network_info
-                })
-
-        for src_net in networks:
-            # we need second cycle to update external object "networks"
-            # with metadata
-            if src_net.get("id") in networks_without_seg_ids:
-                network_info = networks_without_seg_ids[src_net.get("id")]
-                self.create_network(src_net, network_info)
+                # Check segmentation ID for overlapping
+                # If it doesn't overlap with DST, save the same segmentation ID
+                # Otherwise pick free segmentation ID, which does not overlap
+                # with ANY segmentation ID on SRC
+                if seg_id is not None:
+                    # Segmentation ID exists; Check for overlapping
+                    seg_id_overlaps = (network_type in used_seg_ids and
+                                       seg_id in used_seg_ids[network_type])
+
+                    if seg_id_overlaps:
+                        # Choose free segmentation ID, that also does not
+                        # overlap with SRC
+                        src_seg_ids = set(seg_ids[network_type])
+                        dst_seg_ids = set(used_seg_ids[network_type])
+                        busy_seg_ids = src_seg_ids | dst_seg_ids
+
+                        # Search for lowest free segmentation ID
+                        free_seg_id = None
+                        counter = 0
+                        while free_seg_id is None:
+                            if counter not in busy_seg_ids:
+                                free_seg_id = counter
+                            counter += 1
```

Comment:
Would have been much nicer to have this implemented in a separate function/method/class which is covered with unit tests.